### PR TITLE
add spam filter with CAS api https://cas.chat/api

### PR DESCRIPTION
**Pull request by umputun**
this PR adds a check via `https://api.cas.chat/check?user_id=<id>` API on a first message. If the user ID is passed, added to the approved list and won't run a check on that ID anymore

If the user is a spammer, it will be blocked permanently, and the message will be deleted. Dry mode won't remove message and won't ban, just reply with a warning.

### CODE REVIEW

**Comment by umputun** [comment]
@paskal - can you take a look pls?

File: app/main.go [hunk 1]
```diff
@@ -46,6 +46,13 @@ var opts struct {
 	TemplateFile         string           `long:"export-template" default:"logs.html" description:"path to template file"`
 	ExportBroadcastUsers events.SuperUser `long:"broadcast" description:"broadcast-users"`
 
+	SpamFilter struct {
+		Enabled bool          `long:"enabled" env:"ENABLED" description:"enable spam filter"`
+		API     string        `long:"api" env:"CAS_API" default:"https://api.cas.chat" description:"CAS API"`
+		TimeOut time.Duration `long:"timeout" env:"TIMEOUT" default:"5s" description:"CAS timeout in seconds"`
```
Comment (paskal) [inline app/main.go:52]: I think it shouldn't be "in seconds", just "CAS timeout".

File: app/main.go [hunk 2]
```diff
@@ -133,6 +140,12 @@ func main() {
 		openAIBot,
 	}
 
+	if opts.SpamFilter.Enabled {
+		log.Printf("[INFO] spam filter enabled")
```
Comment (paskal) [inline app/main.go:144]: Maybe worth adding API either here or under the debug log entry as we will forget what is it and how it works eventually. Would be great to see it from logs and not only in the code.

File: app/bot/spam_test.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+			dryMode:        false,
+			expectedResp:   Response{},
+		},
+		{
+			name:           "User is a spammer, not in dry mode",
+			mockResp:       `{"ok": false, "description": "Is a spammer"}`,
+			mockStatusCode: 200,
+			dryMode:        false,
+			expectedResp: Response{
+				Text:          "this is spam, go to ban Test User",
```
Comment (paskal) [inline app/bot/spam_test.go:44]:
```suggestion
				Text:          "this is spam, go to ban, Test User",
```
Looks confusing without a comma

File: app/bot/spam.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+		log.Printf("[INFO] user %s is not a spammer, added to aproved", msg.From.Username)
+		s.approvedUsers[msg.From.ID] = true
+		return Response{}
+	}
+
+	log.Printf("[INFO] user %s detected as spammer: %s, msg: %s", msg.From.Username, respData.Description, msg.Text)
+	if s.dry {
+		return Response{Text: "this is spam, but I'm in dry mode, so I'll do nothing yet", Send: true, ReplyTo: msg.ID}
+	}
+	return Response{Text: "this is spam, go to ban " + msg.From.DisplayName, Send: true, ReplyTo: msg.ID, BanInterval: permanentBanDuration, DeleteReplyTo: true}
```
Comment (paskal) [inline app/bot/spam.go:69]:
```suggestion
	return Response{Text: "this is spam, go to ban, " + msg.From.DisplayName, Send: true, ReplyTo: msg.ID, BanInterval: permanentBanDuration, DeleteReplyTo: true}
```

File: app/events/telegram.go [hunk 5]
```diff
@@ -149,6 +150,14 @@ func (l *TelegramListener) Do(ctx context.Context) error {
 				}
 			}
 
+			// delete message if requested by bot
+			if resp.DeleteReplyTo && resp.ReplyTo != 0 {
```
Comment (paskal) [inline app/events/telegram.go:154]: Might be worth the effort adding a test for this case.

**Review by paskal — APPROVED** [review]
LGTM except few small notes.

**Comment by umputun** [comment]
@paskal I have addressed all those except the last one (adding test). It looks like you were the one who added all the tg tests and if you can cover delete too it will be great.

**Comment by paskal** [comment]
Added a test in a separate commit.

**Pull request merged**